src/Controls/src/Core/Page/Page.cs

```diff
@@ -922,5 +922,123 @@ private protected override string GetDebuggerDisplay()
 			var debugText = DebuggerDisplayHelpers.GetDebugText(nameof(BindingContext), BindingContext, nameof(Title), Title);
 			return $"{this.GetType().FullName}: {debugText}";
 		}
+
+#nullable enable
+		IDisposable? _handleNavigatedEventsForRootPage = null!;
+
+		internal void WireUpAsOutgoingPage(Page? outgoingPage, NavigationType navigationType)
+		{
+			_handleNavigatedEventsForRootPage?.Dispose();
+			_handleNavigatedEventsForRootPage = null;
+
+			if (HasNavigatedTo)
+			{
+				SendNavigatingFrom(new NavigatingFromEventArgs());
+			}
+
+			if (IsLoaded)
+			{
+				_handleNavigatedEventsForRootPage =
+					this.OnUnloaded(() =>
+					{
+						if (HasNavigatedTo)
+						{
+							SendNavigatedFrom(new NavigatedFromEventArgs(outgoingPage, navigationType));
+						}
+
+						outgoingPage?.DisconnectHandlers();
+						_handleNavigatedEventsForRootPage?.Dispose();
+						_handleNavigatedEventsForRootPage = null;
+						outgoingPage = null;
+					});
+			}
+			else
+			{
+				if (HasNavigatedTo)
+				{
+					SendNavigatedFrom(new NavigatedFromEventArgs(outgoingPage, navigationType));
+				}
+
+				outgoingPage?.DisconnectHandlers();
+			}
+		}
+
+		internal void WireUpAsIncomingPage(Page? oldPage)
+		{
+			oldPage?.WireUpAsOutgoingPage(oldPage, NavigationType.PageSwap);
```

Comment:
This has `oldPage` both as the object and the parameter. Should this be the same? Will it always be the same? If so, then maybe we can just use `this` in the method and avoid any future confusion.